NDTensors/src/lib/Sectors/src/abstractcategory.jl

```diff
@@ -1,45 +1,130 @@
+  return error("`label_fusion_rule` not defined for type $(category_type).")
+end
+
+# convenient to define fusion rule for LabelledInteger too
+# TBD expose this through ⊗? Currently not accessible.
+function fusion_rule(
+  ::SymmetryStyle, l1::LabelledNumbers.LabelledInteger, l2::LabelledNumbers.LabelledInteger
+)
+  blocks12 = BlockArrays.blocklengths(LabelledNumbers.label(l1) ⊗ LabelledNumbers.label(l2))
+  v =
+    LabelledNumbers.LabelledInteger.(l1 * l2 .* blocks12, LabelledNumbers.label.(blocks12))
+  return GradedAxes.gradedrange(v)
+end
+
+function fusion_rule(
+  ::AbelianGroup, l1::LabelledNumbers.LabelledInteger, l2::LabelledNumbers.LabelledInteger
+)
+  fused = LabelledNumbers.label(l1) ⊗ LabelledNumbers.label(l2)
+  return LabelledNumbers.LabelledInteger(l1 * l2, fused)
+end
+
+# =============  fusion rule and gradedunitrange ===================
+# GradedAxes.tensor_product interface. Only for abelian groups.
+function GradedAxes.fuse_labels(c1::AbstractCategory, c2::AbstractCategory)
+  return GradedAxes.fuse_labels(
+    combine_styles(SymmetryStyle(c1), SymmetryStyle(c2)), c1, c2
+  )
+end
+function GradedAxes.fuse_labels(::SymmetryStyle, c1::AbstractCategory, c2::AbstractCategory)
+  return error("`fuse_labels` is only defined for abelian groups")
+end
+function GradedAxes.fuse_labels(::AbelianGroup, c1::AbstractCategory, c2::AbstractCategory)
+  return fusion_rule(c1, c2)
+end
 
-function dual(category_type::Type{<:AbstractCategory})
-  return error("`dual` not defined for type $(category_type).")
+# cast to range
+to_graded_axis(c::AbstractCategory) = GradedAxes.gradedrange([c => 1])
+to_graded_axis(l::LabelledNumbers.LabelledInteger) = GradedAxes.gradedrange([l])
+to_graded_axis(g::AbstractUnitRange) = g
+
+# allow to fuse a category with a GradedUnitRange
+function GradedAxes.fusion_product(a, b)
+  return GradedAxes.fusion_product(to_graded_axis(a), to_graded_axis(b))
 end
 
-Base.isless(c1::AbstractCategory, c2::AbstractCategory) = isless(label(c1), label(c2))
+# fusion_product with one input to be used in generic fusion_product(Tuple...)
+# TBD define fusion_product() = gradedrange([sector(())=>1])?
+GradedAxes.fusion_product(x) = GradedAxes.fusion_product(to_graded_axis(x))
+
+# product with trivial = easy handling of UnitRangeDual + sort and merge blocks
+GradedAxes.fusion_product(g::AbstractUnitRange) = GradedAxes.fusion_product(trivial(g), g)
+
+function GradedAxes.fusion_product(
+  g1::BlockArrays.BlockedUnitRange, g2::BlockArrays.BlockedUnitRange
+)
+  blocks12 = Vector{eltype(to_graded_axis(fusion_rule(first(g1), first(g2))))}()
+  for l1 in BlockArrays.blocklengths(g1)
+    for l2 in BlockArrays.blocklengths(g2)
+      append!(blocks12, BlockArrays.blocklengths(to_graded_axis(fusion_rule(l1, l2))))
+    end
+  end
+  la3 = LabelledNumbers.label.(blocks12)
+  pairs3 = [r => sum(blocks12[findall(==(r), la3)]; init=0) for r in sort(unique(la3))]
+  out = GradedAxes.gradedrange(pairs3)
+  return out
```

Comment:
If I'm following correctly it looks like this is implemented in two steps:
1. first a "tensor product" is performed by looping over all pairs of blocks and putting the result of fusing each sector into a list, and
2. the blocks generate from 1. are then sorted and merged.

Can we actually implement it in that way using the `tensor_product` function: https://github.com/ogauthe/ITensors.jl/blob/c36654e27db5518b897a941174e24bb7331d9d05/NDTensors/src/lib/GradedAxes/src/fusion.jl#L70-L75? I realize that is written for abelian symmetries right now but it should be easy enough to generalize to non-abelian symmetries, which is essentially what you've done here.

Then we could use `blockmergesort` to sort and merge the blocks. That's how this function was implemented: https://github.com/ITensor/ITensors.jl/blob/v0.6.11/NDTensors/src/lib/GradedAxes/src/abstractgradedunitrange.jl#L147-L152 which I implemented to perform fusion of axes with abelian symmetries, but again it seems like a simple route could be to just use that same kind of design and generalize `tensor_product` to also handle non-abelian symmetries (which I think could be done in a single `tensor_product` implementation by just generalizing `tensor_product` a bit).

Reply:
I propose different conventions for `blockmergesort(dual)`, I implemented the rest.